61 preprocessing operations overlap (#74)
carpentries-incubator/Natural-language-processing: episodes/01-introduction.md

### What is a word?

> When dealing with language our basic data unit is usually a word. We deal with sequences of words and with how they relate to each other to generate meaning in text pieces. Thus, our first step will be to load a text file and provide it with basic structure by splitting it into valid words (this is known as tokenization)!

Reviewer: Somehow this paragraph seems to conflate words and tokens a bit, but I am not sure if it's bad or how to resolve it.

```suggestion
When dealing with language, our basic data unit is usually a word. We deal with sequences of words and with how they relate to each other to generate meaning in text pieces. Thus, our first step will be to load a text file and provide it with basic structure by splitting it into valid words (this is known as tokenization)!
```

71197
```

Splitting by white space is possible but needs several extra steps to get clean words as we know them. We can also use the python `split()` function, which will basically strip any whitespace-like character (including new lines) and get some improvements:

``` python
print("\nProto-Tokens:")
proto_tokens = text.split()
print(proto_tokens[:50])
print(len(proto_tokens))
```

``` output
Proto-Tokens:
['Letter', '1', 'St.', 'Petersburgh,', 'Dec.', '11th,', '17--', 'TO', 'Mrs.', 'Saville,', 'England', 'You', 'will', 'rejoice', 'to', 'hear', 'that', 'no', 'disaster', 'has', 'accompanied', 'the', 'commencement', 'of', 'an', 'enterprise', 'which', 'you', 'have', 'regarded', 'with', 'such', 'evil', 'forebodings.', 'I', 'arrived', 'here', 'yesterday,', 'and', 'my', 'first', 'task', 'is', 'to', 'assure', 'my', 'dear', 'sister', 'of', 'my']
74942
```

however still several extra steps are needed to separate out punctuation appropriately, and perhaps the rules become cumbersome. 

::: callout
### Data Formatting

> Text comes from various sources and are available in different formats (e.g., Microsoft Word documents, PDF documents, ePub files, plain text files, Web pages etc...). The first step is to obtain a clean text representation that can be transferred into python UTF-8 strings that our scripts can manipulate.

Reviewer: Not sure if UTF-8 should be introduced first (or perhaps omitted). Character encoding is a whole issue on its own, especially when you deal with Windows as well.

```suggestion
Texts come from various sources and are available in different formats (e.g., Microsoft Word documents, PDF documents, ePub files, plain text files, Web pages etc.). The first step is to obtain a clean text representation that can be transferred into Python UTF-8 strings that our scripts can manipulate.
```

Data formatting operations might include:

- Removal of special or noisy characters. For example:
    - Random symbols: "The total cost is $120.00#" → remove #

>     - Incorrectly recognized letters or numbers: 1 misread as l, 0 as O, etc. Example: "l0ve" → should be "love"

Reviewer: This refers to recognition through OCR/HTR, right? I think this should be clarified, especially it is being used below.

    - Control or formatting characters: \n, \t, \r appearing in the middle of sentences. Example: "Please\nsubmit\tyour form." → "Please submit your form."

>     - Non-standard Unicode characters: �, �, or other placeholder symbols where OCR failed. Example: "Th� quick brown fox" → "The quick brown fox"

Reviewer: ```suggestion
    - Non-displayable characters: �, �, or other placeholder symbols where OCR failed. Example: "Th� quick brown fox" → "The quick brown fox"
```

- Remove HTML tags (e.g., if you are extracting text from Web pages)
- Strip non-meaningful punctuation (e.g., "The quick brown fox jumps over the lazy dog and con-
tinues to run across the field.)
- Strip footnotes, headers, tables, images etc.
- Remove URLs or phone numbers

And what if you need to extract text from MS Word docs or PDF files or Web pages? There are various Python libraries for helping you extract and manipulate text from these kinds of sources.

- For MS Word documents [python-docx](https://python-docx.readthedocs.io/en/latest/) is popular.

> - For (text-based) PDF files [PyPDF2](https://pypi.org/project/PyPDF2/) and [PyMuPDF](https://pymupdf.readthedocs.io/en/latest/) are widely used. Note that some PDF files are encoded as images (pixels) and not text. If the text in these files is digital (as opposed to scanned handwriting), you can use OCR (Optical Character Recognition) libraries such as [pytesseract](https://pypi.org/project/pytesseract/) to convert the image to machine-readable text.

Reviewer: ```suggestion
- For (text-based) PDF files [PyPDF2](https://pypi.org/project/PyPDF2/) and [PyMuPDF](https://pymupdf.readthedocs.io/en/latest/) are widely used. Note that some PDF files are encoded as images (pixels) and not text. For digitizing printed text, you can use OCR (Optical Character Recognition) libraries such as [pytesseract](https://pypi.org/project/pytesseract/) to convert the image to machine-readable text.
```

- For scraping text from websites, [BeautifulSoup](https://pypi.org/project/beautifulsoup4/) and [Scrapy](https://docs.scrapy.org/en/latest/) are some common options.
- LLMs also have something to offer here, and the field is moving pretty fast. There are some interesting open source LLM-based document parsers and OCR-like extractors such as [Marker](https://github.com/datalab-to/marker), or [PyMuPDF4LLM](https://github.com/pymupdf/PyMuPDF4LLM), just to mention a couple.
:::

### The spaCy NLP Library

> A more sophisticated approach to segment text files is by using specialized NLP libraries. One of the most popular is [spaCy](https://github.com/explosion/spaCy). SpaCy is a free open-source library that focuses on implementing NLP techniques to process text (in several languages, not just English) and extract insights form it in a functional and scalable fashion. Here we will start by using it to segment the text into human-readable tokens. To start, we need to download the pre-trained model, in this case we only need the small English version:

Reviewer: ```suggestion
A more sophisticated approach to segment text files is by using specialized NLP libraries. One of the most popular is [spaCy](https://spacy.io/). SpaCy is a free open-source Python library that focuses on implementing NLP techniques to process text in several languages, and extract insights from it in a functional and scalable fashion. Here we will start by using it to segment the text into human-readable tokens. To start, we need to download a pre-trained model, in this case we only need the small English version:
```

``` python
! python -m spacy download en_core_web_sm
```

This is a model that spaCy already trained for us on a subset of web English data. Hence, the model already "knows" how to obtain clean tokens from English text. When the model processes a string, it does not only do the splitting for us but already provides more advanced linguistic properties of the tokens (such as part-of-speech tags, or named entities). You can check more languages and models in the [spacy documentation](https://spacy.io/models).

Now we will see how spaCy help us to process text and extract interesting properties from it.

### Tokenization 

> Tokenization is a foundational operation in NLP, as it helps to create structure from raw text. This structure is a basic requirement and input for modern NLP algorithms to attribute and interpret meaning from text. This operation involves the segmentation of the text into smaller units referred to as `tokens`. Tokens can be sentences (e.g. `'the happy cat'`), words (`'the', 'happy', 'cat'`), subwords (`'un', 'happiness'`) or characters (`'c','a', 't'`). Different NLP algorithms may require different choices for the token unit. And different languages may require different approaches to identify or segment these tokens.

Reviewer: Should this replace the tokenization introduction above? It seems to overlap at least.

To see how tokenization works using spaCy, let's now import the model and use it to parse our document:

``` python
import spacy

nlp = spacy.load("en_core_web_sm") # we load the small English model for efficiency

doc = nlp(text)
print(type(doc))  # Should be <class 'spacy.tokens.doc.Doc'>
print(len(doc)) # the length of the doc is the number of tokens
print(doc[:50]) # however if we print the doc we get the "raw" text of the first 50 tokens, not the tokens themselves
```

``` output
<class 'spacy.tokens.doc.Doc'>
94553
Letter 1

St. Petersburgh, Dec. 11th, 17--

TO Mrs. Saville, England

You will rejoice to hear that no disaster has accompanied the
commencement of an enterprise which you have regarded with such evil
forebodings.  I arrived here yesterday
```

Now let's access the tokens with spaCy and see what we get: 
```python
# SpaCy-Tokens

> tokens = [token.text for token in doc] # Note that spacy tokens are actually python objects 

Reviewer: ```suggestion
tokens = [token.text for token in doc] # SpaCy tokens are Python objects 
```

print(tokens[:50])
print(len(tokens))
```

``` output
['Letter', '1', '\n\n\n', 'St.', 'Petersburgh', ',', 'Dec.', '11th', ',', '17', '-', '-', '\n\n', 'TO', 'Mrs.', 'Saville', ',', 'England', '\n\n', 'You', 'will', 'rejoice', 'to', 'hear', 'that', 'no', 'disaster', 'has', 'accompanied', 'the', '\n', 'commencement', 'of', 'an', 'enterprise', 'which', 'you', 'have', 'regarded', 'with', 'such', 'evil', '\n', 'forebodings', '.', ' ', 'I', 'arrived', 'here', 'yesterday']
94553
```

::: callout
A good word tokenizer for example, does not simply break up a text based on spaces and punctuation, but it should be able to distinguish:

-   abbreviations that include points (e.g.: *e.g.*)
-   times (*11:15*) and dates written in various formats (*01/01/2024* or *01-01-2024*)
-   word contractions such as *don't*, these should be split into *do* and *n't*
-   URLs

> Many older tokenizers are rule-based, meaning that they iterate over a number of predefined rules to split the text into tokens, which is useful for splitting text into word tokens for example. Modern large language models use subword tokenization, which learn to break text into pieces that are statistically convenient, this makes them more flexible but less human-readable.

Reviewer: ```suggestion
Many older tokenizers are rule-based, meaning that they iterate over a number of predefined rules to split the text into tokens, which is useful for splitting text into word tokens for example. Modern language models use subword tokenization, which break text into pieces according to statistical patterns, which are often not aligned with meaning and human intuition.
```

:::

The differences look subtle at the beginning, but if we carefully inspect the way spaCy splits the text, we can see the advantage of using a specialized tokenizer. 

::: callout
We do not have to depend necessarily on the `Doc` and `Token` spaCy objects. Once we tokenized the text with a spaCy model, we can extract the list of words as a list of strings and continue our text analysis:

```python
# List of string tokens
token_list = [token.text for token in doc if "\n" not in token.text]
print(token_list[:50])

# Or a tokenized (each word separated by a single space) string of the whole text

> tokenized_text = " ".join(token_list)

Reviewer: It's unrelated, but I would like to use best practices of naming variable along the way, including *not* encoding the type of a variable in its name.

```suggestion
tokens = [token.text for token in doc if "\n" not in token.text]
print(tokens[:50])

# Or a tokenized (each word separated by a single space) string of the whole text
tokenized_text = " ".join(tokens)
```

Author: Interesting, here I think `text` does not refer to the type of the variable but to the "original text", anyway I tried to avoid `tokens` in order to not confuse the token objects with the text property of the token. Will think of a better term

EDIT: only now I see you meant the `token_list`, hence you are right!

Reviewer: I agree, I would keep `tokenized_text` as-is, only change `token_list` to `tokens` (per my suggested change). This also allows changing the type from list to e.g. generator without having to rename the variable.

3317
```

> Note that in this case each sentence is a python object, and the property `.text` returns an untokenized string (in terms of words). But we can still access the list of word tokens inside each sentence object if we want:

Reviewer: ```suggestion
Note that in this case each sentence is a Python object, and the property `.text` returns an untokenized string (in terms of words). But we can still access the list of word tokens inside each sentence object if we want:
```

```python
sents_sample = list(doc.sents)[:10]
for sent in sents_sample:
    print("Sentence:", sent.text)
    for token in sent:
        print("\tToken:", token.text)
```

### Lowercasing

Removing uppercases to e.g. avoid treating "Dog" and "dog" as two different words could also be useful, for example to train word vector representations, where we want to merge both occurrences as they represent exactly the same concept. Lowercasing can be done with Python directly as:

```python
lower_text = text_flat.lower()
lower_text[:100] # Beware that this is a python string operation
```

Beware that lowercasing the whole string as a first step might affect the tokenizer behavior since tokenization benefits from information provided by case-sensitive strings. We can therefore tokenize first using spaCy and then obtain the lowercase strings of each token using the `.lower_` property:

```python
lower_text = [token.lower_ for token in doc]
lower_text[:10] # Beware that this is a list of strings now!
```

### Lemmatization 

> Another way to normalize words in a text is to transform them into their *dictionary form*. Consider how "eating", "ate", "eaten" are all variations of the root verb "eat". Each variation is sometimes known as an _inflection_ of the root word. Conversely, we say that the word "eat" is the _lemma_ for the words "eating", "eats", "eaten", "ate" etc. Lemmatization is therefore the process of rewriting each token or word in a given input text as its lemma. You can also use lemmatization for generating word embeddings. For example, you can have a single vector for `eat` instead of one vector per verb tense. 

Reviewer: ```suggestion
Another way to normalize words in a text is to transform them into their *dictionary form*. Consider how "eating", "ate", "eaten" are all variations of the root verb "eat". Each variation is known as an _inflection_ of the root word. Conversely, we say that the word "eat" is the _lemma_ for the words "eating", "eats", "eaten", "ate" etc. Lemmatization is therefore the process of rewriting each word in a given input text as its lemma. You can also use lemmatization for generating word embeddings. For example, you can have a single vector for `eat` instead of one vector per verb tense. 
```

Lemmatization is not only a possible preprocessing step in NLP but also an NLP task on its own, with different algorithms for it. Therefore we also tend to use pre-trained models to perform lemmatization. Using spaCy we can access the lemmmatized version of each token with the `lemma_` property (notice the underscore!):

```python
lemmas = [token.lemma_ for token in doc]
print(lemmas[:50])
```

Note that the list of lemmas is now a list of strings.

### Named Entities

> The spaCy pipeline already runs by default more advanced task, such as Named Entity Recognition (NER), the task of identifying words or prhases that refer to unique real-world instances (normally proper nouns). You can access the entities with:

Reviewer: ```suggestion
The default spaCy pipeline already runs more advanced tasks, such as Named Entity Recognition (NER), the task of identifying words or phrases that refer to unique real-world instances (normally proper nouns). You can access the entities with:
```
